pilot/ARCHITECTURE.md

```diff
@@ -0,0 +1,163 @@
+#### Partial Computations
+Along with caching, partial computations are a critical performance optimization to ensure that we do not need to build (or send) every resource to every proxy on every change. This is discussed more in the Config Serving section.
+
+### Config Serving
+Config serving is the layer that actually accepts proxy clients, connected over bidirectional gRPC streams, and serve them the required configuration.
+
+We will have two triggers for sending config - requests and pushes.
+
+#### Requests
+
+Requests come from the client specifically asking for a set of resources. This could be requesting the initial set of resources on a new connection, or from a new dependency. For example, a push of `Cluster X` referencing `Endpoint Y` may lead to a request for `Endpoint Y` if it is not already known to the client.
+
+Note that clients can actually send three types of messages - requests, ACKs of previous pushes, and NACKs of previous pushes. Unfortunately, these are not clearly distinguished in the API, so there is some logic to split these out (`shouldRespond`).
+
+#### Pushes
+
+A push occurs when Istiod detects an update of some set of configuration is needed. This results in roughly the same result as a Request (new configuration is pushed to the client), and is just triggered by a different source.
+
+Various components described in Config Ingestion can trigger a Config Update. These are batched up ("debounced"), to avoid excessive activity when many changes happen in succession, and eventually enqueued in the Push Queue.
+
+The Push Queue is mostly a normal queue, but it has some special logic to merge push requests for each given proxy. This results in each proxy having 0 or 1 outstanding push requests; if additional updates come in the existing push request is just expanded.
+
+Another job polls this queue and triggers each client to start a push.
+
+```mermaid
+graph TD
+	subgraph Config Flow
+		cu(Config Update)
+		db(Debounce)
+		pc(Recompute Push Context)
+		pq(Push Queue)
+		cu-->db
+		db--Trigger Once Steady-->pc
+		pc--Enqueue All Clients-->pq
+	end
+	
+	subgraph Proxy
+		c(Client)
+	end
+	subgraph Pusher
+		pj(Push Job)
+		pj--read-->pq
+		pj--trigger-->c
+	end
+```
+At a high level, each client job will find the correct generator for the request, generate the required configuration, and send it.
+#### Optimizations
+A naive implementation would simply regenerate all resources, of all subscribed types, for each client, on any configuration change. However, this scales poorly. As a result, we have many levels of optimizations to avoid doing this work.
+
+First, we have a concept of a `Full` push. Only `Full` pushes will recompute `PushContext` on change; otherwise this is skipped and the last `PushContext` is re-used. Note: even when `Full`, we try to copy as much from the previous `PushContext` as possible. For example, if only a `WasmPlugin` changed, we would not recompute services indexes.
```

Comment:
What determines/decides when a 'Full' push is necessary, and (roughly) how?

Reply:
Its basically "did I check the entire codebase and make sure that all consumers of this event support Full=false (some generators skip it entirely) AND that push-context does not use it". Not great...